### PR TITLE
Adjust AutoLoader

### PR DESCRIPTION
**Pull request by tyleruebele**
Now uses a file based class map
Now builds a dependency graph in that classmap

### CODE REVIEW

File: src/AutoLoader.php [hunk 1]
```diff
@@ -290,4 +354,239 @@ public static function isNamespaceKnown(string $className) {
...
+        if (preg_match('{(<\?)(?!(php|hh))}i', $contents)) {
+            $contents = preg_replace('{//.* | /\*(?:[^*]++|\*(?!/))*\*/}x', '', $contents);
+        }
+        preg_match_all('{
+            (?:
+                 \b(?<![\$:>])(?P<type>class|interface|trait) \s++ (?P<name>[a-zA-Z_\x7f-\xff:][a-zA-Z0-9_\x7f-\xff:\-]*+)
+                 (?:\s+extends\s+(?P<extends>[\\\\a-zA-Z_\x7f-\xff:][\\\\a-zA-Z0-9_\x7f-\xff:\-]*+))?
+                 (?:\s+implements\s+(?:(?P<implements>[\\\\a-zA-Z_\x7f-\xff:][\\\\a-zA-Z0-9_\x7f-\xff:\-]*+)(?:,\s*)?)+)?
+               | \b(?<![\$:>])(?P<ns>namespace) (?P<nsname>\s++[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+(?:\s*+\\\\\s*+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+)*+)? \s*+ [\{;]
+               | \b(?<![\$:>])(?P<use>use) \s++ (?P<usename>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+(?:\s*+\\\\\s*+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+)*+) (?: \s*+ as \s*+ (?P<usealias>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+))? \s*+;
```
Comment (markneumann99) [inline src/AutoLoader.php]:
```suggestion
               | \b(?<![\$:>])(?P<use>use) \s++ (?P<usename>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+(?:\s*+\\\\\s*+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+)*+) (?: \s++ as \s++ (?P<usealias>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+))? \s++;
```

File: src/AutoLoader.php [hunk 2]
```diff
@@ -290,4 +354,239 @@ public static function isNamespaceKnown(string $className) {
...
+        // strip comments if short open tags are in the file
+        if (preg_match('{(<\?)(?!(php|hh))}i', $contents)) {
+            $contents = preg_replace('{//.* | /\*(?:[^*]++|\*(?!/))*\*/}x', '', $contents);
+        }
+        preg_match_all('{
+            (?:
+                 \b(?<![\$:>])(?P<type>class|interface|trait) \s++ (?P<name>[a-zA-Z_\x7f-\xff:][a-zA-Z0-9_\x7f-\xff:\-]*+)
+                 (?:\s+extends\s+(?P<extends>[\\\\a-zA-Z_\x7f-\xff:][\\\\a-zA-Z0-9_\x7f-\xff:\-]*+))?
+                 (?:\s+implements\s+(?:(?P<implements>[\\\\a-zA-Z_\x7f-\xff:][\\\\a-zA-Z0-9_\x7f-\xff:\-]*+)(?:,\s*)?)+)?
+               | \b(?<![\$:>])(?P<ns>namespace) (?P<nsname>\s++[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+(?:\s*+\\\\\s*+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+)*+)? \s*+ [\{;]
```
Comment (markneumann99) [inline src/AutoLoader.php]:
```suggestion
               | \b(?<![\$:>])(?P<ns>namespace) \s++ (?P<nsname>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+(?:\s*+\\\\\s*+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+)*+)? \s*+ [\{;]
```

File: src/AutoLoader.php (hunk 1 above)
Comment (tyleruebele) [reply src/AutoLoader.php]: The spaces around `as` should be adjusted as indicated.
The space before the `;` should remain optional.

File: src/AutoLoader.php (hunk 2 above)
Comment (tyleruebele) [reply src/AutoLoader.php]: This seems to work either way, but I made the adjustment.

**Review by markneumann99 — APPROVED** [review]

**Pull request merged**